### PR TITLE
fix(web): support for notany() and context()

### LINKED ISSUE #917: notany() and context() do not work well together in KeymanWeb
The `notany()` statement and the `context()` statement do not cooperate in generated KeymanWeb code. Currently, a `context` statement that references the `notany()` in the context will return a compile error. This dramatically limits the usefulness of `notany()` ... because there is no way to capture the character that was matched. 

For example:

```
store(&TARGETS) 'web'

begin Unicode > use(main)

group(main) using keys

store(vow) 'aeiou'
notany(vow) 'b' + 'c' > context(1) 'BC'
```

I did some investigation into possible solutions. This is a partial fix:

```
diff --git a/windows/src/developer/TIKE/compile/CompileKeymanWeb.pas b/windows/src/developer/TIKE/compile/CompileKeymanWeb.pas
index d45c8ca3..d923ee5c 100644
--- a/windows/src/developer/TIKE/compile/CompileKeymanWeb.pas
+++ b/windows/src/developer/TIKE/compile/CompileKeymanWeb.pas
@@ -746,6 +746,7 @@ var
   pwszcontext,pwsz: PWideChar;
   Index: Integer;   // I3910
   nlt: string;
+  ContextLen: Integer;
 
   function AdjustIndex(pwszContext: PWideChar; Index: Integer): Integer;   // I3910
   var
@@ -785,6 +786,11 @@ var
           end;
         CODE_DEADKEY:
           Result := Result + nlt + Format('k.KDO(%d,t,%d);', [len, recContext.Deadkey.Deadkey]);   // I4611
+        CODE_NOTANY:
+          begin
+            Index := ContextLen - ContextIndex + 1; // Need to get context index from cached context, based on length of context
+            Result := Result + nlt + Format('k.KO(%d,t,k.KC(%d,1,t));', [len, Index]);
+          end;
         else
         begin
           ReportError(fkp.Line, CERR_NotSupportedInKeymanWebContext, Format('Statement %s is not currently supported in CODE_CONTEXT match', [GetCodeName(recContext.Code)]));  // I1971   // I4061
@@ -815,6 +821,8 @@ begin
 
   if Assigned(fkp) then
   begin
+    ContextLen := xstrlen(fkp.dpContext);
+
     if IsKeyboardVersion10OrLater
     // KMW >=10.0 use the full, sentinel-based length for context deletions.
     then
```

However, it does not cope with start-of-context checks (although a naive perusal seems to indicate that it should). It also has not been tested against the 10.0 code paths. A complete spec needs to be written to validate the requirements and compatibility.

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #917.

This PR goes along with #3816 to enable support for `notany()` and `context()` references from the output.

### CODE REVIEW

File: common/core/web/keyboard-processor/src/text/kbdInterface.ts [hunk 1]
```diff
@@ -457,7 +465,7 @@ namespace com.keyman.text {
                   deadContext[i].set();
                 }
                 // 'n' for 'notany'.  If we actually match or if we have nul context (\uFFFE), notany fails.
-              } else if(r.n && (result || context[i] !== NUL_CONTEXT)) {
+              } else if(r.n && (result || context[i] === NUL_CONTEXT)) {
```
Comment (mcdurdin) [inline common/core/web/keyboard-processor/src/text/kbdInterface.ts:468]: This test seemed to be backwards. It could be dealt with as a separate PR because it's a separate bug, but I'm reasonably comfortable with rolling it in with this fix.

I feel like this change needs a unit test, but wanted to put this in front of you first.

File: web/testing/issue917-context-and-notany/index.html [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+        <tr><td>adcd</td><td>'a' DK(1) 'XX'</td><td>Second position in context, match deadkey rule on line 27</td></tr>
+        <tr><td>adcd?</td><td>'pass!'</td><td>Verifies that deadkey is conserved through rule match on line 27</td></tr>
+        <tr><td>axcd</td><td>'axXX'</td><td>Verifies that match occurs for character on line 27</td></tr>
+        <tr><td>axcd?</td><td>'axXX?'</td><td>Sanity check for line 27</td></tr>
+
+        <tr><td>fghi</td><td>'fgXX'</td><td>Testing adjusted index for `if` statement on line 35</td></tr>
+        <tr><td>fghi?</td><td>'fgXX?'</td><td>Sanity check for line 35</td></tr>
+        <tr><td>fdhi</td><td>'f' DK(1) XX'</td><td>Testing deadkey match for adjusted index for `if` statement on line 35</td></tr>
+        <tr><td>fdhi?</td><td>'pass!'</td><td>Verifies that deadkey is conserved thrghou rule match on line 35</td></tr>
```
Comment (mcdurdin) [inline web/testing/issue917-context-and-notany/index.html:68]: Manual tests that could be converted into automatic tests...

File: common/core/web/keyboard-processor/src/text/kbdInterface.ts [hunk 3]
```diff
@@ -787,7 +795,36 @@ namespace com.keyman.text {
       outputTarget.restoreProperties();
     }
   
-    
+    /**
+     * `contextExOutput` function emits the character or object at `contextOffset` from the
+     * current matched rule's context. Introduced in Keyman 14.0, in order to resolve a
+     * gap between desktop and web core functionality for context(n) matching on notany().
+     * See #917 for additional detail.
+     * @alias       KNO
```
Comment (jahorton) [inline common/core/web/keyboard-processor/src/text/kbdInterface.ts:803]: Why `KNO`?  Is it for **n**otany**Output** or something like that?  Most of the shorthand/abbreviated forms are relatively simple to interpret, especially when referring to the TS-source 'long names'... but this one eludes me a bit.  I'd prefer something that matches a little more closely.

File: common/core/web/keyboard-processor/src/text/kbdInterface.ts [hunk 4]
```diff
@@ -787,7 +795,36 @@ namespace com.keyman.text {
...
+        this.output(Pdn, outputTarget, "");
+      }
+
+      const context = this.ruleContextEx.get(contextLength, contextLength);
+      const dk = context.deadContext[contextOffset-1], vc = context.valContext[contextOffset-1];
+      if(dk) {
+        outputTarget.insertDeadkeyBeforeCaret(dk.d);
+      } else if(typeof vc == 'string') {
+        this.output(-1, outputTarget, vc);
+      } else {
```
Comment (jahorton) [inline common/core/web/keyboard-processor/src/text/kbdInterface.ts:823]: :+1:

File: common/core/web/keyboard-processor/src/text/kbdInterface.ts (hunk 1 above)
Comment (jahorton) [reply common/core/web/keyboard-processor/src/text/kbdInterface.ts:468]: Seems fine to me; it's related enough.

File: common/core/web/keyboard-processor/src/text/kbdInterface.ts (hunk 3 above)
Comment (mcdurdin) [reply common/core/web/keyboard-processor/src/text/kbdInterface.ts:803]: > Is it for **n**otany**O**utput or something like that?

Yep. But feel free to suggest a better short name 😁. `KCXO`?

File: web/testing/issue917-context-and-notany/index.html (hunk 2 above)
Comment (jahorton) [reply web/testing/issue917-context-and-notany/index.html:68]: The most applicable site for this would be found at the following path:  [`common/core/web/keyboard-processor/tests/cases/engine/context.js`](https://github.com/keymanapp/keyman/blob/master/common/core/web/keyboard-processor/tests/cases/engine/context.js).

The test keyboard should be placed within `common/core/web/tests/resources/keyboards`, as that's where a number of common web test resources are stored.  (In particular, the resources already referenced by the tests within the site mentioned above.)

The patterns for existing unit tests within that file are quite well-structured, so I don't expect much difficulty with a conversion.  Feel free to either include them within that file or write a separate but similarly structured file.  Alternatively, I should be able to do the conversion once I'm in a good place to spare a couple of hours.

**Comment by mcdurdin** [comment]
One thing I couldn't quite figure out was why the `CachedContextEx` object was a triple-nested array. It seems like overkill (that, and the split of `Deadkey` and `Val` into two separate arrays which is also somewhat awkward). Can you eludicate on the design -- is it inheriting some existing legacy cruft or am I missing something obvious?

(In either case, I'd like to add a comment regarding the structure for Mr Posterity)

File: web/testing/issue917-context-and-notany/index.html (hunk 2 above)
Comment (jahorton) [reply web/testing/issue917-context-and-notany/index.html:68]: Probably something separate, as the actual rule application (`runEngineRuleSet`) defined in that file is very tailored to `fullContextMatch`.  Aside from that, and possibly some details of the test spec pattern, the rest of the file should be pretty applicable.

For programatically emulating keystrokes (as seen within `runEngineRuleSet`):

```
      var matchTest = matchDefs[j];
      var ruleSeq = new KMWRecorder.RecordedKeystrokeSequence(matchTest.sequence);
      let proctor = new KMWRecorder.NodeProctor(keyboard, device, assert.equal);

      // We want to specify the OutputTarget for this test; our actual concern is the resulting context.
      var target = new com.keyman.text.Mock();
      ruleSeq.test(proctor, target);
```

The final line in that block executes the keystroke sequence defined by each spec's `.sequence` property.  In fact, if all you want to do is check against equality with the intended output, that's the core part you'd need; everything afterward was established for extra-thorough checks against `fullContextMatch`, which won't matter for you.  Just the `sequence` definitions should suffice for automating these tests.

**Comment by jahorton** [comment]
> One thing I couldn't quite figure out was why the `CachedContextEx` object was a triple-nested array. It seems like overkill (that, and the split of `Deadkey` and `Val` into two separate arrays which is also somewhat awkward). Can you eludicate on the design -- is it inheriting some existing legacy cruft or am I missing something obvious?
> 
> (In either case, I'd like to add a comment regarding the structure for Mr Posterity)

`cachedContextEx[n][ln][index]`:
- `n` - Given the caret's current position, the number of characters to shift our lookup's starting position to the left
- `ln` - the length of the string to return, starting at the position marked at `n`.
- `index` - an index within the returned substring.


Compare against plain `CachedContext` - _that_ one is definitely "existing legacy cruft", so I'm pretty sure that `CachedContextEx` simply modeled itself after that, uh, "already-existing evil".

---

Noting that `CachedContext` is designed for direct `string` comparison, its pattern makes sense; it avoids the need to recompute substrings for multiple comparisons.

`CachedContextEx`, on the other hand... isn't really designed for direct comparisons, true.  We probably could drop the `ln` index entirely with ease... if it weren't for one of the extra checks contained within the `fullContextMatch` test case file.  (The check asserts that the constructed live cache result deep-matches a preknown value.)  There's a chance some `fullContextMatch` logic may depend on the behavior as well.   

tl;dr: with time investment, we could absolutely drop the `ln` index.  But it would require a bit of time investment.

File: web/testing/issue917-context-and-notany/index.html (hunk 2 above)
Comment (mcdurdin) [reply web/testing/issue917-context-and-notany/index.html:68]: Tests created in notany_context.js 💯

File: common/core/web/keyboard-processor/src/text/kbdInterface.ts (hunk 3 above)
Comment (jahorton) [reply common/core/web/keyboard-processor/src/text/kbdInterface.ts:803]: `KCXO` sounds fine... I'm fine with `X` representing `notany`, as a whole representing _**c**_ontext _**(x)**_notany _**o**_utput.

Comment (darcywong00) [reply common/core/web/keyboard-processor/src/text/kbdInterface.ts:803]: So `KCXO` is to get used in #3816 and added to help.keyman-staging.com/developer/engine/web/14.0/reference/interface ?

**Review by jahorton — APPROVED** [review]
